pkg/provisioner/ironic/bios.go

```diff
@@ -0,0 +1,203 @@
+package ironic
+
+import (
+	"fmt"
+	"reflect"
+	"sort"
+	"strconv"
+
+	"github.com/gophercloud/gophercloud/openstack/baremetal/v1/nodes"
+
+	metal3v1alpha1 "github.com/metal3-io/baremetal-operator/apis/metal3.io/v1alpha1"
+)
+
+// BuildBIOSCleanSteps build clean steps for different driver
+func BuildBIOSCleanSteps(driver string, firmware *metal3v1alpha1.FirmwareConfig) (cleanSteps []nodes.CleanStep, err error) {
+	// Deal possible panic
+	defer func() {
+		r := recover()
+		if r != nil {
+			err = fmt.Errorf("panic in build BIOS settings: %v", r)
+		}
+	}()
+
+	switch driver {
+	case "ilo", "ilo5":
+		cleanSteps = buildILOCleanSteps(firmware)
```

Comment:
We have a whole package of vendor-specific modules for the BMCs to avoid having similar logic in the provisioner. If we make the BMC responsible for building the clean steps, or create a similar but different package, we can avoid using reflection and avoid having driver or vendor specific logic here.

Reply:
Thanks, I add a new function(`BuildBIOSCleanSteps`) to interface `AccessDetails`.